src/librustc/infer/anon_types/mod.rs

```diff
@@ -489,11 +504,33 @@ struct Instantiator<'a, 'gcx: 'tcx, 'tcx: 'a> {
 impl<'a, 'gcx, 'tcx> Instantiator<'a, 'gcx, 'tcx> {
     fn instantiate_anon_types_in_map<T: TypeFoldable<'tcx>>(&mut self, value: &T) -> T {
         debug!("instantiate_anon_types_in_map(value={:?})", value);
+        let tcx = self.infcx.tcx;
         value.fold_with(&mut BottomUpFolder {
-            tcx: self.infcx.tcx,
-            fldop: |ty| if let ty::TyAnon(def_id, substs) = ty.sty {
-                self.fold_anon_ty(ty, def_id, substs)
-            } else {
+            tcx,
+            fldop: |ty| {
+                if let ty::TyAnon(def_id, substs) = ty.sty {
+                    // Check that this is `impl Trait` type is declared by
+                    // `parent_def_id`. During the first phase of type-check, this
+                    // is true, but during NLL type-check, we sometimes encounter
+                    // `impl Trait` types in e.g. inferred closure signatures that
+                    // are not 'local' to the current function and hence which
```

Comment:
This comment is confusing.

If we had `typeof` or `impl Trait` associated types, this could just as well happen within the signature of a function. The thing is, we want to make a distinction between "syntactic" `impl Trait` types that are "bound to our function" and "free" `impl Trait` types that are not. 

Aka if we have this:
```Rust
mod some_mod {
    type Foo = impl Iterator; // not sure about the exact syntax
}

mod my_mod {
    fn foo(x: ::some_mod::Foo) -> impl Iterator { x }
}
```

Then in `foo`, `::some_mod::Foo` is a "free" `TyAnon` while the return type is a "bound" `TyAnon`. Luckily for us, uses of `TyAnon` can't be recursive (for now?), so if a `TyAnon` has our own `DefId` then it is definitely bound to our function.

I'm not sure this is the best way to express this, but that's how things are set up for now and we are not going to refactor the world for NLL.

Reply:
In the case you give, we are still returning an `impl Iterator`, and it would still be local to `foo()` -- that is, when this function executes, we would have effectively desugared to:

```rust
mod my_mod {
    abstract type FooReturn: Iterator;
    fn foo(x: ::some_mod::Foo) -> FooReturn { x }
```

and hence we would be "instantiating" `FooReturn` (whose value would then get inferred to a `TyAnon` representing `some_mod::Foo`).

In fact, this is basically exactly the case that this code is designed to handle in NLL: you can construct a very similar scenario with closures and return type inference today.

> I'm not sure this is the best way to express this, but that's how things are set up for now and we are not going to refactor the world for NLL.

I'm not really sure what this quite means. It doesn't seem like NLL requires any particular refactoring of the world here -- it's just that NLL exposes us to more complex cases than we had to handle before, but it's not like it's a problem to handle them. I can try to reword the comment, anyhow.

Reply:
Pushed a second attempt! =)